src/arraytypes/dictencoding.jl

```diff
@@ -114,19 +114,29 @@ function arrowvector(::DictEncodedType, x, i, nl, fi, de, ded, meta; dictencode:
         # dict encoding doesn't exist yet, so create for 1st time
         if DataAPI.refarray(x) === x
```

Comment:
Note that there's no guaranty that `refarray` returns an array of contiguous integer indices, nor that `refpool` is defined when `refarray` is. To make this more general it would make sense to check that, and if that's not the case then create a `PooledArray` as in the first branch.

Reply:
I don't think we're assuming that `refarray` returns contiguous integers, because we're still doing `inds = copyto!(similar(Vector{signed(eltype(refa))}, length(refa)), refa)`, which should still work if non-contiguous. 

I guess we could add the check `|| refpool(x) === nothing` here. Curious, what are cases where `refarray` is defined, but not `refpool`?

Reply:
> I don't think we're assuming that `refarray` returns contiguous integers, because we're still doing `inds = copyto!(similar(Vector{signed(eltype(refa))}, length(refa)), refa)`, which should still work if non-contiguous.

At least you assume `signed(eltype(refa))` is defined, but the docstring for `refarray` doesn't say that. Currently I'm not aware of types for which that's not the case, but better be safe. (I have no idea about whether contiguity is required.)

> I guess we could add the check `|| refpool(x) === nothing` here. Curious, what are cases where `refarray` is defined, but not `refpool`?

Well again I'm just following the docstrings. Historically, we introduced `refarray` and `refvalue` before `refpool`, so clearly implementing the two former doesn't imply implementing the latter. But conversely, when you implement `refpool` you have to implement `refarray`, otherwise you can't satisfy requirements from its docstring. I think the asymmetry makes sense since in general, implementing `refpool` can be harder as it requires knowing the set of possible values in advance.